### PR TITLE
8337939: ZGC: Make assertions and checks less convoluted and explicit

### PR DESCRIPTION
**Pull request by jsikstro**
There are currently cases where calls to type converters are made only to assert whether the conversion is reasonable or not and then discarding the result. For example, to_zaddress(...) is used to check if the pointer passed to it is a valid zaddress or not, whilst discarding the result of the conversion.

Additionally, a call like oopDesc::is_oop(to_oop(o)) is convoluted since a similar check to is_oop() is already done inside to_oop(), which should be a separate operation in its entirety.

Asserts/checks in affected places should be separated so that assertion/checking can be explicitly made and not done more than necessary.

Tested with tiers 1-7 on linux64 and linux64-debug.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8337939](https://bugs.openjdk.org/browse/JDK-8337939): ZGC: Make assertions and checks less convoluted and explicit (**Enhancement** - P4)

### Reviewers
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**) 🔄 Re-review required (review applies to [42044a86](https://git.openjdk.org/jdk/pull/20478/files/42044a8697e048bba874721a2d2e2371ead6a2fa))
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**) 🔄 Re-review required (review applies to [42044a86](https://git.openjdk.org/jdk/pull/20478/files/42044a8697e048bba874721a2d2e2371ead6a2fa))

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20478/head:pull/20478` \
`$ git checkout pull/20478`

Update a local copy of the PR: \
`$ git checkout pull/20478` \
`$ git pull https://git.openjdk.org/jdk.git pull/20478/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20478`

View PR using the GUI difftool: \
`$ git pr show -t 20478`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20478.diff">https://git.openjdk.org/jdk/pull/20478.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20478#issuecomment-2273563616)

### CODE REVIEW

File: src/hotspot/share/gc/z/zVerify.cpp [hunk 1]
```diff
@@ -111,6 +111,16 @@ static bool z_is_null_relaxed(zpointer o) {
...
 
+static void z_verify_oop_object(zaddress addr, zpointer o, void* p) {
+  const oop obj = cast_to_oop(addr);
+  guarantee(oopDesc::is_oop(obj), BAD_OOP_ARG(o, p));
+}
+
+static void z_verify_root_oop_object(zaddress o, void* p) {
+  const oop obj = cast_to_oop(o);
+  guarantee(oopDesc::is_oop(obj), BAD_OOP_ARG(o, p));
+}
```
Comment (stefank) [inline src/hotspot/share/gc/z/zVerify.cpp:122]: I pre-reviewed this part, but I realize now that I'd like to update the parameter name for the zaddress. Would you mind updating the code this?
```suggestion
static void z_verify_root_oop_object(zaddress addr, void* p) {
  const oop obj = cast_to_oop(addr);
  guarantee(oopDesc::is_oop(obj), BAD_OOP_ARG(addr, p));
}
```

**Review by stefank — CHANGES_REQUESTED** [review]

**Review by stefank — APPROVED** [review]
Looks good.

**Review by albertnetymk — APPROVED** [review]

**Review by fisk — APPROVED** [review]
Looks good.

**Review by stefank — APPROVED** [review]

**Pull request closed without merging**